overlord/boot/firstboot.go

```diff
@@ -154,26 +177,31 @@ func importAssertionsFromSeed(st *state.State) error {
 
 	// collect
 	var modelAssertion *asserts.Model
-	assertionsToAdd := make([]asserts.Assertion, len(dc))
+	assertionsToAdd := make([]asserts.Assertion, 0, len(dc))
 	bs := asserts.NewMemoryBackstore()
-	for i, fi := range dc {
-		content, err := ioutil.ReadFile(filepath.Join(assertSeedDir, fi.Name()))
-		if err != nil {
-			return fmt.Errorf("cannot read assertion: %s", err)
-		}
-		as, err := asserts.Decode(content)
+	for _, fi := range dc {
+		fn := filepath.Join(assertSeedDir, fi.Name())
+		assertions, err := readAsserts(fn)
 		if err != nil {
-			return fmt.Errorf("cannot decode assertion: %s", err)
+			return fmt.Errorf("cannot read assertions: %s", err)
 		}
-		if err := bs.Put(as.Type(), as); err != nil {
-			return err
-		}
-		assertionsToAdd[i] = as
-		if as.Type() == asserts.ModelType {
-			if modelAssertion != nil {
-				return fmt.Errorf("cannot add more than one model assertion")
+		for _, as := range assertions {
+			if err := bs.Put(as.Type(), as); err != nil {
+				if revErr, ok := err.(*asserts.RevisionError); ok {
+					if revErr.Current >= as.Revision() {
+						// we already got something more recent
+						continue
+					}
+				}
+				return err
+			}
+			assertionsToAdd = append(assertionsToAdd, as)
+			if as.Type() == asserts.ModelType {
+				if modelAssertion != nil {
+					return fmt.Errorf("cannot add more than one model assertion")
```

Comment:
Nice that the logic ought to work even if we have two model assertions, as long as they have matching keys, due to the revision logic above.